### PR TITLE
Add predictive address functionality to recipient field in send

### LINKED ISSUE #1165: Add predictive autocomplete to address input component
We should store every address a person ever sends to or has into an array that's used to suggest addresses to a user when they're typing.  This will be even more valuable once we add [ENS support](https://github.com/MetaMask/metamask-plugin/issues/1130).

### PR DESCRIPTION
**Pull request by 2-am-zzz**
Fixes #1165 by adding auto-complete functionality to addresses! It even works with .eth resolvers! Yay!
screenshot from 2017-03-09 15-12-56

Needs discussion is:
- How should we determine which addresses stay in the address book long-term?
- Styling issues (the arrow to show the entire address book and the .eth icon overlap)
- How many addresses to show?

### CODE REVIEW

**Comment by kumavis** [comment]
This is really awesome 😸

**Comment by flyswatter** [comment]
My asks:
- Make the account names to be unicode and not bold.
- Sort the suggestions by either recently used, most used, or ideally, a combination.
- Show a max of like 5-10 addresses.

**Comment by Zanibas** [comment]
@flyswatter 

+ datalists are rendered on a browser-by-browser basis, and it looks like styling for it is limited at the moment.
+ It doesn't seem like suggestions can be sorted...according to my research. At least, with the default datalist component. Doing the two above would require an undertaking to recreate our own custom datalist.
+ Should be doable, but is related to recently used. I can write the code in such a way that can't control the _order_, but it can control the _content_ if I include some additional data with each entry.

**Comment by flyswatter** [comment]
- If styling is limited, can we at least make the addresses not bold?  Maybe make the names bold instead?
- Ok, skip the sorting, I guess the list will already be limited to matches.
- I guess this isn't that important either, but we *should* have a max number of addresses that we retain, just to guard against heavy usage.

**Comment by kumavis** [comment]
@flyswatter @Zanibas can we merge this feature and then add style polish?

**Comment by Zanibas** [comment]
@kumavis I'd be good with that--the only stylings I could find are ones that involve polyfill...

File: ui/app/components/ens-input.js [hunk 1]
```diff
@@ -46,6 +47,28 @@ EnsInput.prototype.render = function () {
     style: { width: '100%' },
   }, [
     h('input.large-input', opts),
+    // The address book functionality.
+    h('datalist',
```
Comment (kumavis) [inline ui/app/components/ens-input.js]: you can do this with `datalist#addresses` and omit the params obj

Comment (Zanibas) [inline ui/app/components/ens-input.js]: Done!

**Comment by kumavis** [comment]
yeah lame that styling it is so limited...
maybe we can omit `label` and leave `value` as the name
and then resolve the name before processing
that should at least look better

**Comment by flyswatter** [comment]
What's this look like now?

**Comment by flyswatter** [comment]
Ok, I pulled this down, built it, looked at the code, and now I see what you mean by datalist is picky.  Hopefully we can find some refinements eventually, but for now, this is a really cool feature.

**Review by danfinlay — APPROVED** [review]

**Pull request merged**